### PR TITLE
Scroll through every object in a class using an id cursor

### PR DESCRIPTION
**Pull request by antas-marcin**
### What's being changed:

Adds support for Scroll API.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: usecases/objects/get.go [hunk 1]
```diff
@@ -134,15 +135,21 @@ func (m *Manager) getObjectFromRepo(ctx context.Context, class string, id strfmt
...
-	sort, order *string, additional additional.Properties,
+func (m *Manager) getObjectsFromRepo(ctx context.Context,
+	offset, limit *int64, sort, order *string, after *string,
+	additional additional.Properties,
 ) ([]*models.Object, error) {
 	smartOffset, smartLimit, err := m.localOffsetLimit(offset, limit)
 	if err != nil {
 		return nil, NewErrInternal("list objects: %v", err)
 	}
+	// TODO: after
```
Comment (aliszka) [inline usecases/objects/get.go:146]: ?

File: usecases/objects/get.go [hunk 2]
```diff
@@ -235,3 +242,10 @@ func (m *Manager) trackUsageList(res search.Results) {
 	}
 	m.metrics.AddUsageDimensions(res[0].ClassName, "get_rest", "list_include_vector", res[0].Dims)
 }
+
+func (m *Manager) getScroll(after *string, offset, limit *int64) (*filters.Scroll, error) {
```
Comment (aliszka) [inline usecases/objects/get.go:246]: what is the `offset` for?

**Comment by etiennedi** [comment]
I've converted this from a draft to a regular one. This is mainly because anything that's in draft mode right now will be discarded for v1.18, but this should make it in, of course.

File: adapters/repos/db/shard_read.go [hunk 3]
```diff
@@ -312,6 +315,35 @@ func (s *Shard) allObjectList(ctx context.Context, limit int,
...
+	if err != nil {
+		return nil, errors.Wrap(err, "after argument")
+	}
+	out := make([]*storobj.Object, scroll.Limit)
+
+	i := -1
+	cursor := s.store.Bucket(helpers.ObjectsBucketLSM).Cursor()
+	defer cursor.Close()
+
+	for k, v := cursor.Seek(idBytes); k != nil && i < scroll.Limit; k, v = cursor.Next() {
```
[the remark this answers is not in the document]
Comment (etiennedi) [reply adapters/repos/db/shard_read.go]: > next in order after non existing

This.

File: adapters/repos/db/crud_integration_test.go [hunk 4]
```diff
@@ -39,7 +39,7 @@ import (
 	"github.com/weaviate/weaviate/usecases/replica"
 )
 
-func TestCRUD(t *testing.T) {
+func TestCRUD_(t *testing.T) {
```
Comment (aliszka) [inline adapters/repos/db/crud_integration_test.go:42]: why name change?

File: adapters/repos/db/shard_read.go [hunk 5]
```diff
@@ -312,6 +315,50 @@ func (s *Shard) allObjectList(ctx context.Context, limit int,
...
+	defer cursor.Close()
+
+	var key, val []byte
+	if len(idBytes) == 0 {
+		// start from the beginning
+		key, val = cursor.First()
+	} else {
+		key, val = cursor.Seek(idBytes)
+		seekID, err := uuid.FromBytes(key)
+		if err == nil && seekID.String() == scroll.After {
```
Comment (aliszka) [inline adapters/repos/db/shard_read.go:343]: no `err != nil` handling

File: usecases/traverser/explorer_validate_scroll.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+	"github.com/weaviate/weaviate/entities/dto"
+	"github.com/weaviate/weaviate/entities/filters"
+	"github.com/weaviate/weaviate/entities/schema"
+)
+
+func (e *Explorer) validateScroll(params dto.GetParams) error {
+	if params.Scroll != nil {
+		if params.Group != nil || params.HybridSearch != nil || params.KeywordRanking != nil ||
+			params.NearObject != nil || params.NearVector != nil || len(params.ModuleParams) > 0 {
```
Comment (aliszka) [inline usecases/traverser/explorer_validate_scroll.go:25]: shouldn't there be `nearText` somewhere?

**Review by aliszka — COMMENTED** [review]
I don't see test where `limit` is tested as well. scroll seems to be used always in loop until last element - that way it is not clear whether single call returns all remaining elements or just up to given `limit`

File: usecases/traverser/explorer_validate_scroll.go (hunk 6 above)
Comment (etiennedi) [reply usecases/traverser/explorer_validate_scroll.go:25]: len of `len(params.ModuleParams)` will be `>0` if `nearText` is set, as `nearText` is only present on a setup with a `t2v` module.

Comment (antas-marcin) [reply usecases/traverser/explorer_validate_scroll.go:25]: correct, `len(params.ModuleParams) > 0` covers this

**Pull request merged**